packages/xchain-haven/src/haven/sync-handler.ts

```diff
@@ -0,0 +1,95 @@
+import { getAddressInfo } from './api'
+import { SyncObserver, SyncStats } from './types'
+import { assertIsDefined } from './utils'
+
+const SYNC_THRESHOLD = 300
+
+class SyncHandler {
+  private scannedHeight: number | undefined = undefined
+  private blockHeight: number | undefined = undefined
+  private updateSyncProgressIntervalID: ReturnType<typeof setInterval> | undefined
+  private syncObservers: SyncObserver[] = []
+
+  async subscribeSyncProgress(newObserver: SyncObserver) {
+    // do check if we need to track a sync progress at all
+    if ((await this.isSyncing()) && this.updateSyncProgressIntervalID === undefined) {
+      this.updateSyncProgressIntervalID = setInterval(() => this.updateSyncProgress(), 10 * 1000)
+    }
+
+    this.syncObservers.push(newObserver)
+    const syncState = this.getSyncState()
+    this.updateSyncObservers([newObserver], syncState)
+  }
+
+  async updateSyncProgress() {
+    await this.fetchSyncProgress()
+
+    assertIsDefined(this.scannedHeight)
+    assertIsDefined(this.blockHeight)
+
+    const syncState: SyncStats = await this.getSyncState()
+    this.updateSyncObservers(this.syncObservers, syncState)
+
+    if (!(await this.isSyncing())) {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      //@ts-ignore
+      clearInterval(this.updateSyncProgressIntervalID) // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```

Comment:
No need to ignore/disable eslint: `this.updateSyncProgressIntervalID` might be undefined, just add a check before (similar you have done in `purge`)